### PR TITLE
[stdlib] Modernize Set and Dictionary internals

### PR DESCRIPTION
**Pull request by lorentey**
Clean up Set and Dictionary internals to improve readability and to prepare for upcoming resiliency improvements.

- Nest public iterator types under the corresponding collection type, with a compatibility typealias.
    `SetIterator<Element>` ⟹ `Set<Element>.Iterator`
    `DictionaryIterator<Key, Value>` ⟹ `Dictionary<Key, Value>.Iterator`

- Nest auxiliary types under the corresponding higher-level type. Remove confusing `Buffer` suffix. Unify naming of enums representing native/cocoa variants.
    `_VariantSetBuffer<Element>` ⟹ `Set<Element>._Variant`
    `SetIndexRepresentation<Element>` ⟹ `Set<Element>.Index._Variant`
    `SetIteratorRepresentation<Element>` ⟹ `Set<Element>.Iterator._Variant`

    `_NativeSetBuffer<Element>` ⟹ `_NativeSet<Element>`
    `_NativeSetIndex<Element>` ⟹ `_NativeSet<Element>.Index`

    `_CocoaSetBuffer` ⟹ `_CocoaSet`
    `_CocoaSetIndex` ⟹ `_CocoaSet.Index`
    `_CocoaSetIterator` ⟹ `_CocoaSet.Iterator`

    (Ditto for `Dictionary`.)

- Add explicit types representing native iterators: `_NativeSet<Element>.Iterator` and `_NativeDictionary<Key, Value>.Iterator`. Previously these were exploded into associated values on the native case of the top-level iterator variant.

- Make Cocoa helper class names a little bit more helpful in Cocoa context.
    `_NativeSetNSEnumerator` ⟹ `_SwiftSetNSEnumerator`
    `_NativeDictionaryNSEnumerator` ⟹ `_SwiftDictionaryNSEnumerator`

- Adjust internal property and variable names accordingly.
- To enhance readability, remove internal typealiases wherever possible.
- Audit inlinability.

rdar://problem/42585384

### CODE REVIEW

File: stdlib/public/core/Set.swift [hunk 1]
```diff
@@ -390,7 +382,7 @@ extension Set: Collection {
   /// to the set. Don't expect any particular ordering of set elements.
   ///
   /// If the set is empty, the value of this property is `nil`.
-  @inlinable // FIXME(sil-serialize-all)
+  @inlinable
   public var first: Element? {
```
Comment (airspeedswift) [inline stdlib/public/core/Set.swift]: I wonder if this really does materially perform better than the default.

Comment (lorentey) [reply stdlib/public/core/Set.swift]: The default creates an iterator, which complicates the native case, but improves the bridged case -- `_CocoaSet.startIndex` is currently O(n), while the iterator uses fast enumeration. 😦

Ideally this should dispatch to `_variant` and do the right thing in each case. Or we could just remove it.

Comment (lorentey) [reply stdlib/public/core/Set.swift]: Actually, it turns out this is the correct way to implement this, despite the performance hit. 

Bridged indices are *currently* based on `-[NSDictionary allKeys]`, which is not guaranteed to use the same element ordering as an enumerator. (Which indicates Dictionary/Set iterators may sometimes return elements in a different order than indices. 😭)

I think we should change indices to use enumerators, and soon.

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
The benchmarks shouldn't be affected, although there were a couple changes to inlinability.
@swift-ci please benchmark

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
The `DictionaryRemove` benchmark regression looks legit: the compiler decided not to inline some `_UnsafeBitmap` members. Bitmap operations should always be inlined; I expect that enforcing this will fix the regression.

**Comment by lorentey** [comment]
The plot thickens!

**Comment by lorentey** [comment]
Please test with following PR:
https://github.com/apple/swift-lldb/pull/797

@swift-ci please test

**Comment by lorentey** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

**Comment by lorentey** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

**Comment by lorentey** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

**Comment by lorentey** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

**Comment by lorentey** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

**Comment by lorentey** [comment]
@swift-ci please test and merge

**Comment by lorentey** [comment]
@swift-ci please test and merge

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by lorentey** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
That first wall time regression was just noise, then, not a problem.

**Pull request merged**